### PR TITLE
getTimeZoneOffset: fix offset bug for old dates (#3924)

### LINKED ISSUE #3924: Invalid date conversion for old dates in local tz
# 🐛 Bug Report 

When using `parseAbsolute(value, "Europe/Oslo")` as a value in `useDateFieldState`, an odd bug occurs.
The minutes/seconds/dates are being updated on update to year.

See [minimal example](https://codesandbox.io/s/funny-brattain-i87tq4?file=/src/DateInput.js)

## 🤔 Expected Behavior

Changing the year does not affect other fields.

## 😯 Current Behavior

Updating year also changes minute/second/day (depending on granularity).

## 💁 Possible Solution

It seems to me that `parseAbsolute` returns wrong timezoneOffset for old dates, in certain timezones
```ts
parseAbsolute('0002-01-01T00:00:00Z', 'Europe/Oslo')
// returns { ... , offset :3180000 }}, but should maybe be 3212000?
```

### Workaround

Using a `CalendarDateTime` rather than `ZonedDateTime`. See method `workaroundParseDate` in [`DateInput.js`](https://codesandbox.io/s/funny-brattain-i87tq4?file=/src/DateInput.js)

## 🔦 Context

Trying to use @react-aria/datepicker, with timezone `Europe/Oslo`.

## 💻 Code Sample

See [minimal example](https://codesandbox.io/s/funny-brattain-i87tq4?file=/src/DateInput.js)

## 🌍 Your Environment

I suppose @internationalized/date is the culprit.

| Software         | Version(s) |
| ---------------- | ---------- |
| [@internationalized/date](https://codesandbox.io/examples/package/@internationalized/date) | 3.0.2 |
| [@react-aria/datepicker](https://codesandbox.io/examples/package/@react-aria/datepicker) | 3.2.1 |
| [@react-aria/i18n](https://codesandbox.io/examples/package/@react-aria/i18n) | 3.6.3 |
| Browser          | Chrome, Firefox, Safari
| Operating System | Mac OSX 12.1

### PR DESCRIPTION
**Pull request by tomfa**
Fixes issue where `getTimeZoneOffset` would not include seconds offset (applicable for old dates) when system time zone equalled passed time zone.

Closes #3924 

## ✅ Pull Request Checklist:

- [x] Included link to corresponding [React Spectrum GitHub Issue](https://github.com/adobe/react-spectrum/issues).
- [ ] ~Added/updated unit tests and storybook for this change (for new code or code which already has tests).~  
   Not sure how to unit test this one – it would have to change system `TZ`
- [x] Filled out test instructions.
- [ ] ~Updated documentation (if it already exists for this component).~ (Not applicable)
- [ ] ~Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/TR/wai-aria-practices-1.1/)~ (Not applicable)

## 📝 Test Instructions:

- Set browser timezone to `Europe/Oslo`
- Use [this codebox example]( https://codesandbox.io/s/happy-wozniak-838vce)
- See the bug
- Update code as in this PR (Search for `Fast path: for local timezone` in `node_modules/@internationalized`)
- See bug not appearing (anymore)

### CODE REVIEW

**Comment by tomfa** [comment]
(Signed the CLA now)

**Comment by snowystinger** [comment]
Going to close and reopen so the CLA kicks in

**Comment by snowystinger** [comment]
Thanks so much for the PR, any chance you'd be willing to write a unit test for this?

**Comment by tomfa** [comment]
> Thanks so much for the PR, any chance you'd be willing to write a unit test for this?

Here's how I imagine one could write a test.

- Mock `getLocalTimeZone` to return a timezone where the bug occurs.
- Mock `new Date().getTimezoneOffset()` to return an fake offset (bug doesn't occur with env `TZ=UTC`, which I presume tests are run in).
- Compare `getTimeZoneOffset` against a test implementation of `getTimeZoneOffset` without the fast path.

I'm not sure that really brings any value. If you have suggestions on other ways to test it that is more meaningful, I'm all ears :)

**Comment by snowystinger** [comment]
Maybe I'm not quite following the bug. Here are my thoughts.

We actually completely mock the clock in most of our tests, jest.useFakeTimers https://jestjs.io/docs/jest-object#fake-timers

So I don't think we'd need to compare getTimeZoneOffset against getTimeZoneOffset without the fast path, so much as we could write the test with predetermined date/times https://jestjs.io/docs/jest-object#jestsetsystemtimenow-number--date and compare against what we know the result to be for that predetermined date/time

I think you are correct we may need to mock getLocaltTimeZone

**Comment by tomfa** [comment]
The bug occurs on a line that uses `Date.getTimezoneOffset()`. To add a test, we would probably want to control the result of that method. 

If I'm not mistaken, the result of `Date.getTimezoneOffset()` depends on the system running jest tests (can be overridden with env `TZ=UTC`) and is not mockable with `jest.useFakeTimers`.

I'm not sure how to add a test in an elegant way, because:
- Changing env var may affect other tests.
- Changing `Date.prototype.getTimezoneOffset` may affect other tests when tests run in parallell.
- `jest.useFakeTimers` doesn't handle this case.

I would be ❤️ very happy ❤️  to be mistaken on any of the above, and learn how to add a good test for this. But I suggest you (or someone else) write that test?

[Here's a few good test cases](https://github.com/moment/moment/issues/5515#issuecomment-692609688) for dates/timezones using second offsets

**Comment by tomfa** [comment]
I can add a skipped test case with a comment on how it should be run manually for documrntation purposes?

**Comment by LFDanLu** [comment]
> I can add a skipped test case with a comment on how it should be run manually for documentation purposes?

Would that essentially be the test instructions you've included for this PR? If so, then I think those test instructions good enough.

> I'm not sure how to add a test in an elegant way, because:
Changing env var may affect other tests.
Changing Date.prototype.getTimezoneOffset may affect other tests when tests run in parallel.
jest.useFakeTimers doesn't handle this case.

Before I dive in to investigate mocking the env var for the timezone and seeing what possible breaks that occur, did you run into any of the above issues when trying to implement the test or is the above the concerns you had for any possible test written for this bug?

**Comment by tomfa** [comment]
Added a test now (using the unfortunate hacks) @LFDanLu

File: packages/@internationalized/date/tests/conversion.test.js [hunk 1]
```diff
@@ -138,6 +139,35 @@ describe('CalendarDate conversion', function () {
...
+          return s;
+        });
+
+        const tzOffset = getTimeZoneOffset(new Date(date).getTime(), timezone);
+        expect(tzOffset).toBe(expectedOffset);
+
+        jest.clearAllMocks();
+        clearLocalTimeZoneCache();
+      }
+      delete process.env.TZ;
```
Comment (snowystinger) [inline packages/@internationalized/date/tests/conversion.test.js:167]: Can't we just do something like this? or what am I overlooking? it succeeds with your changes and fails without your change.

```suggestion
      for (let [timezone, date, expectedOffset] of testCases) {
        jest.spyOn(Intl.DateTimeFormat.prototype, 'resolvedOptions').mockImplementation(function () {
          let s = resolvedOptions.call(this);
          s.timeZone = timezone;
          return s;
        });

        const tzOffset = getTimeZoneOffset(new Date(date).getTime(), timezone);
        expect(tzOffset).toBe(expectedOffset);

        jest.clearAllMocks();
      }
```

Comment (tomfa) [reply packages/@internationalized/date/tests/conversion.test.js]: I updated the test according to your suggestions :+1:

I come short at communicating the problems with this. But I agree that testing this edge case has some value as documentation.

- If you want to discover the issues, you can add test cases for 2 different timezones (with date > 1970).
- If you want to discuss how this can be tested, you can send me an email to set up a video chat 😄

**Review by snowystinger — APPROVED** [review]
